### PR TITLE
Making theory order local per dataset

### PR DESCRIPTION
**Pull request by jacoterh**
This PR removes the `pQCD` order key in the runcard and replaces it by locally specifying it per dataset, e.g:

```
datasets:

  ATLAS_CMS_SSinc_RunI: {"order": "NLO_QCD"}
  ATLAS_CMS_tt_AC_8TeV: {"order": "NLO_QCD"}
  ATLAS_SSinc_RunII: {"order": "NLO_QCD"}
  ATLAS_STXS_runII_13TeV: {"order": "NLO_QCD"}
  ATLAS_WH_Hbb_13TeV: {"order": "NLO_QCD"}
```

The reason we need this feature is because we now have different higher order corrections that we like to distinguish: `NLO_QCD, NLO_EW, NLO_QCD_EW`. The theory files need updating accordingly, which is done in [PR#25](https://github.com/LHCfitNikhef/smefit_database/pull/25) on smefit_database.

### CODE REVIEW

**Comment by jacoterh** [comment]
@LucaMantani we probably need to merge this first and then rebase #96 onto main.

**Comment by LucaMantani** [comment]
@jacoterh Correct me if wrong, this will be fully compatible with current tables as long as we specify "NLO" or "LO" for each dataset, right?

**Comment by jacoterh** [comment]
Yes, that's right, the only thing that mattes is that the order specified for each dataset also appears in the theory json. This can be any key as you point out, but PR#25 on smefit_database actually updates the keys in the jsons to our new convention (i.e. NLO_QCD, NLO_EW, etc)

File: tests/test_fisher.py [hunk 1]
```diff
@@ -71,9 +71,8 @@ def test_fisher():
         operators_to_keep = np.array(["Op1", "Op2", "Op3"])
         dataset = load_datasets(
             commondata_path,
-            datasets=["data_test5"],
+            datasets={"data_test5": {"order": "NLO"}},
```
Comment (arossia94) [inline tests/test_fisher.py:74]: Long-term idea: a simple converter of the runcards from the old format to the new one could be useful.

**Review by arossia94 — CHANGES_REQUESTED** [review]
Minor requests regarding default and limit behaviors.
Should we also update the documentation within this same PR before forgetting about this change?

**Comment by jacoterh** [comment]
Should be ready to go. Updated behaviour

```
datasets:
  - dataset_1
  - dataset_2: {"order": "NLO_QCD"}
  - dataset_3: {"order": "NLO_EW"}
```
where the predictions of `dataset_1` are LO (default behaviour),  `dataset_2` NLO_QCD and  `dataset_3` at NLO_EW

**Comment by jacoterh** [comment]
Updated syntax (no mixed types anymore)

```
datasets:
  - name: dataset_1
  - name: dataset_2
    order: NLO_QCD
  - name: dataset_3
    order: "NLO_EW
```

This syntax is future proof as it supports naturally additional properties we might want to specify per dataset, i.e. cuts.

**Comment by LucaMantani** [comment]
> Updated syntax (no mixed types anymore)
> 
> ```
> datasets:
>   - name: dataset_1
>   - name: dataset_2
>     order: NLO_QCD
>   - name: dataset_3
>     order: "NLO_EW
> ```
> 
> This syntax is future proof as it supports naturally additional properties we might want to specify per dataset, i.e. cuts.

So now you made it a list of dictionaries, right? it's more similar to the NNPDF format. I am in favour of this but I thought we wanted to keep the functioning of the old runcards?

**Comment by jacoterh** [comment]
That's right, it's a list of dictionaries, so it'd be equivalent to doing

```
datasets:
  - {name: dataset_1}
  - {name: dataset_2, order: NLO_QCD}
  - {name: dataset_3, order: "NLO_EW
```
This breaks compatibility with the old runcards, but for this we can provide a short script to convert them. In any case it's a small difference.

File: src/smefit/loader.py [hunk 2]
```diff
@@ -602,18 +602,16 @@ def load_datasets(
...
     Loader.commondata_path = pathlib.Path(commondata_path)
-    if theory_path is not None:
-        Loader.theory_path = pathlib.Path(theory_path)
-    else:
-        Loader.theory_path = pathlib.Path(commondata_path)
+    Loader.theory_path = pathlib.Path(theory_path or commondata_path)
 
     _logger.info(f"Applying cutoff scale: {cutoff_scale} GeV.")
-    for sset in np.unique(datasets):
+    for sset in datasets:
```
Comment (LucaMantani) [inline src/smefit/loader.py:608]: Here and in similar loops (the one in load_rge_mat for example) datasets are looped without being put in alphabetical order now? I am a bit worried about having mismatches, are we guaranteed that things are ordered correctly?

**Review by LucaMantani — COMMENTED** [review]
I have one comment on the order datasets are looped through but for the rest it looks fine.

File: src/smefit/loader.py (hunk 2 above)
Comment (jacoterh) [reply src/smefit/loader.py:608]: I see your concern, but datasets is a list which has a well defined ordering, so there's no need to sort anything (as opposed to dictionaries). But it was a list also before, and yet we had it got sorted. Perhaps @giacomomagni can comment if/why this was necessary at the time?

**Review by LucaMantani — APPROVED** [review]

File: src/smefit/loader.py [hunk 3]
```diff
@@ -545,11 +545,11 @@ def load_datasets(
     commondata_path,
     datasets,
     operators_to_keep,
-    order,
     use_quad,
     use_theory_covmat,
     use_t0,
     use_multiplicative_prescription,
+    default_order="LO",
```
Comment (giacomomagni) [inline src/smefit/loader.py:552]: why do you need to have a default "LO" ?  Do you need to load a dataset without any theory being loaded?

File: src/smefit/loader.py [hunk 4]
```diff
@@ -564,12 +564,12 @@ def load_datasets(
     ----------
         commondata_path : str, pathlib.Path
             path to commondata folder, commondata excluded
-        datasets : list
-            list of datasets to be loaded
+        datasets : List
```
Comment (giacomomagni) [inline src/smefit/loader.py:567]:
```suggestion
        datasets : list
```

File: src/smefit/loader.py (hunk 2 above)
Comment (giacomomagni) [reply src/smefit/loader.py:608]: I guess the unique was there just to be sure a dataset is not loaded twice, but yes from the `DataTuple` you can always infer the order.

**Review by arossia94 — APPROVED** [review]

File: src/smefit/loader.py (hunk 3 above)
Comment (jacoterh) [reply src/smefit/loader.py:552]: This is just the default theory order that gets loaded if one doesn't specify it for a particular dataset. Of course, the default can be set to NLO as well, but not all predictions are available at NLO.

Comment (giacomomagni) [reply src/smefit/loader.py:552]: sure, but sorry to be picky why do you want to have a default ? 
I mean if you load, you need to specify an order

**Pull request merged**